libnmstate/nm/connection.py

```diff
@@ -182,241 +189,265 @@ def get_setting_duplicate(self, setting_name):
         return setting
 
     def _active_connection_callback(self, src_object, result, user_data):
-        cancellable = user_data
-        self._mainloop.drop_cancellable(cancellable)
+        if self._ctx.is_cancelled():
+            self._activation_clean_up()
+            return
+        action = user_data
 
         try:
             nm_act_con = src_object.activate_connection_finish(result)
         except Exception as e:
-            act_type, act_object = self._get_activation_metadata()
-
-            if self._mainloop.is_action_canceled(e):
-                logging.debug(
-                    "Connection activation canceled on %s %s: error=%s",
-                    act_type,
-                    act_object,
-                    e,
-                )
-            elif self._is_connection_unavailable(e):
-                logging.warning(
-                    "Connection unavailable on %s %s, retrying",
-                    act_type,
-                    act_object,
-                )
-                self._reset_profile()
-                time.sleep(0.1)
-                self.safe_activate_async()
-            else:
-                self._mainloop.quit(
-                    "Connection activation failed on {} {}: error={}".format(
-                        act_type, act_object, e
-                    )
-                )
+            self._ctx.fail(NmstateLibnmError(f"{action} failed: error={e}"))
             return
 
         if nm_act_con is None:
-            act_type, act_object = self._get_activation_metadata()
-            self._mainloop.quit(
-                "Connection activation failed on {} {}: error=unknown".format(
-                    act_type, act_object
+            self._ctx.fail(
+                NmstateLibnmError(
+                    f"{action} failed: "
+                    "error='None return from activate_connection_finish()'"
                 )
             )
         else:
-            devname = nm_act_con.props.connection.get_interface_name()
+            devname = self.devname
```

Comment:
Originally we intentionally re-fetched the iface name to make sure it did not changed (I think). Why now it is not needed? Because we are no longer re-trying to activate?

Reply:
The devname should never change during activation for all current supported actions.

No sure what retry means here. We wait on state change signal on NM.Device and NM.ActivateConnection.